spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/inbound/PubSubInboundChannelAdapter.java

```diff
@@ -51,10 +50,10 @@
 
 	private AckMode ackMode = AckMode.AUTO;
 
-	private MessageConverter messageConverter;
-
 	private HeaderMapper<Map<String, String>> headerMapper = new PubSubHeaderMapper();
 
+	private InboundMessageMapper<byte[]> messageConverter;
```

Comment:
Let's call it `messageMapper` then!

OTOH I won't mind if we evolve our `PubSubMessageConverter` abstraction and will really reuse it here as well, even if there is nothing to do with the `PubSubTemplate` here in the `PubSubInboundChannelAdapter`.

At least that is how we do in other places like Kafka or AMQP, or even JMS.

That's really embarrassing from my side that I led you to that generics `MessageConverter`  in the beginning: i just wanted to reuse an existing abstraction as much as possible without extra code to support here, but as we all realized already it doesn't fit our model. So, let's try to look into some other existing (hope working) solution in the mentioned projects!

Reply:
I think we can just add support in the `PubSubTemplate` for something more convenient than 
`com.google.cloud.pubsub.v1.MessageReceiver`, something that has 

```
void receiveMessage(Object payload, Map<String,String> headers, AckReplyConsumer consumer);
```
rather than
```
void receiveMessage(final PubsubMessage message, final AckReplyConsumer consumer);
```
This way we can remove conversion concerns from the inbound side as well.

Reply:
??? That's something not related. `PubSubTemplate`  is for single, user-driven operations. Meanwhile this `PubSubInboundChannelAdapter` is really message-driven, based on events generated by the `Consumer` in the background thread. So, it isn't going to be so convenient to inject `PubSubTemplate` just for the conversion reasons.

Reply:
I'm not sure I understand. `PubSubInboundChannelAdapter` already has the `PubSubTemplate` injected, and it relies on the `MessageReceiver`.
```java
	@Override
	protected void doStart() {
		super.doStart();

		this.subscriber =
				this.pubSubTemplate.subscribe(this.subscriptionName, this::receiveMessage);
	}

	private void receiveMessage(PubsubMessage pubsubMessage, AckReplyConsumer consumer) {
...
```

I'm just suggesting an enhancement to the `PubSubTemplate` that can also be taken advantage of in the adapter.

Reply:
Oh! Sorry. I really missed something in this project.
So, I see now that `pubSubTemplate` has that `subscribe()` operation. A bit unusual for the stateless component when the *template* abstraction should be just only for a use-driven operations, but let it be - you broke the approach and you are in responsibility to support it.
That wasn't my call to pull `subscribe` to the `pubSubTemplate`: https://github.com/spring-cloud/spring-cloud-gcp/pull/148

Nevertheless I disagree with the `receiveMessage(Object payload, Map<String,String> headers)` API. You still have a `HeaderMapper` in the Channel Adapter. Are you going to move that one to the template one day as well  ? 🤷‍♂️ 

For the conversion part you really can call a getter for converter from that template and still be good in the Channel Adapter.

With all of these discussions and massive reworks I feel like you would like to get rid of Spring Integration support at all and you try to pull as much of its functionality into one component as possible. Meanwhile we still should keep in mind the level of responsibility and loosely coupling concern.

Reply:
I believe in the early discussions there was a drive to do all the hard work in the template, and have the channel adapters reuse the template as much as possible. I agree that may well violate the loose coupling and separation of concerns.

I really just want to keep message conversion in one place, be it the template or the channel adapters. If you think it belongs best in the channel adapters, then let's take it out of the template along with convenience methods like 
```java
	@Override
	public <T> ListenableFuture<String> publish(String topic, T payload) {
		return publish(topic, payload, null);
	}
```

where you can pass an arbitrary object rather than the `PubsubMessage`.
If you think that's better, I'm OK with it. I was never really a supporter of these convenience methods, and rather wanted to keep the template as minimal as possible.

Reply:
Well, actually that's fine to keep those convenient method. As long as we have similar API in Spring AMQP and Spring Kafka.

My only point is to get a `PubSubMessageConverter` from the template for conversion in the Channel Adapter since such a `subscribe()` is really out of template scope and that your `receive()` suggestion is a bit confusing. 

Well, it really confuses me more and more since template should expose only atomic stateless operations.

With the  `PubSubMessageConverter`  abstraction you have just introduced we really have a one-stop-shop for conversion which can be reused in the template as well as in the Channel Adapters.

Let's for now reuse it via getter from the template to complete this issue and we can revise the template architecture a bit later!